enhancements/sdk-supports-operands-deletion.md

```diff
@@ -0,0 +1,135 @@
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The purpose of this enhancement is to outline how the [Operator SDK (OSDK)][sdk] can support operator developers targeting [Operator Lifecycle Manager (OLM)][olm] and need to clean up operands prior to an uninstall via custom resource (CR) deletion. OSDK will provide mechanisms to enable and facilitate CR deletion for operators that are installed using OLM and operator developers can explicitly opt-in for the cleanup of their operator workloads.
+
+## Motivation
+
+An operand is the program underlying a CR that an operator controls.
+From this proposal’s perspective a CR controlled by the operator is the only resource object _directly_ cleaned up by OLM. Other resources such as pods, volumes, and off-cluster resources created by the operator need to be cleaned up by the operator.
+
+In kubernetes garbage collection model, an object can be an owner of another object within the same namespace. These "dependent" objects have an [OwnerReference][owner-reference] field that references the "parent" object. This relationship can be specified manually by the user, or automatically for certain resources on creation. If user deletes the parent object, Kubernetes garbage collector will delete all of its dependent objects (sometimes automatically if it is specified).
+
+Cross-namespace owner references are not allowed by design; instead, a [finalizer][finalizers] should be used to perform cleanup. Finalizers are more powerful than owner references because they permit a controller to define how an object is cleaned up.
+
+In OLM's [operand cleanup proposal][operand-cleanup-proposal] to handle operand deletion, OLM would configure the cleanup API in the [ClusterServiceVersion (CSV)][csv] and deleting the CSV will be the trigger to initiate operand cleanup. The CSV spec would have an optional field `spec.cleanup` which lets the user configure the intent to cleanup operands before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+Upon setting `spec.cleanup.enabled: true`, OLM will add `operatorframework.io/cleanup-apis` finalizer on the CSV to prevent operator removal before cleanup. See [operand cleanup proposal][operand-cleanup-proposal] for more details.
+
+## Goals
+
+- Provide default scaffolding for an operand cleanup
+- Cleanup is not opted in by default to preserve current behavior
+- `operator-sdk cleanup` should optionally delete CRDs, and use OLM's operand cleanup feature
+
+## Possible Future Goals (Not in scope)
+
+- SDK should have a mechanism to handle events/alerts received by OLM in case cleanup is blocked
+- The timeout for operand deletion should be user configurable by the SDK
```

Comment:
It's implementable but we may not do this in the first pass, just to avoid the complexity that comes with it. 

If it is user configurable, the user/operator author needs to be aware of how long the cleanup process would take and in some cases they may not know what value it would be.

SDK could enforce a generic default timeout for all operators but that would mean that for a long running cleanup process, we might prematurely conclude that it's stuck and try to unblock operand deletion and potentially leave things lingering. So for now, we will not implement timeouts, and it will be one of the risks.